### PR TITLE
Rationalize vpc in specs

### LINKED ISSUE #504: Rationalize specification of VPC and subnet in resource specs
Specification of locations should include…

* `vpcName` e.g. "vpc0" for clusters / security groups / load balancers
* `subnetPurpose` e.g. "internal" for clusters / load balancers

We can then determine the VPC id (which we need to send to Orca) from those.

### PR DESCRIPTION
**Pull request by robfletcher**
I was bothered by the way the specification of VPC was inconsistent between resource types. Would love to know if this stab at making it more sensible is legit.

Potentially fixes #504

### CODE REVIEW

File: keel-api/src/test/resources/examples/alb-example.yml [hunk 1]
```diff
@@ -4,14 +4,14 @@ kind: application-load-balancer
...
 spec:
-  vpcName: vpc0
   subnetType: internal (vpc0)
   moniker:
     app: keeldemo
     stack: example
     detail: albec2v1
   locations:
     accountName: test
+    vpcName: vpc0
```
Comment (robfletcher) [inline keel-api/src/test/resources/examples/alb-example.yml]: Generally moving `vpcName` from the root of the spec into the `locations`

File: keel-api/src/test/resources/examples/cluster-example.yml [hunk 2]
```diff
@@ -14,6 +14,7 @@ spec:
       type: DEB
   locations:
     accountName: test
+    vpcName: vpc0
```
Comment (robfletcher) [inline keel-api/src/test/resources/examples/cluster-example.yml]: I ended up adding this to cluster for consistency. Not sure if that's desirable

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/Location.kt [hunk 3]
```diff
@@ -20,6 +20,7 @@ package com.netflix.spinnaker.keel.api.ec2
 data class Location(
   val accountName: String,
   val region: String,
-  val subnet: String?, // TODO: is this actually optional?
+  val vpcName: String,
+  val subnet: String,
```
Comment (robfletcher) [inline keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/Location.kt]: VPC and subnet now required

File: keel-api/src/test/resources/examples/alb-example.yml [hunk 4]
```diff
@@ -4,14 +4,13 @@ kind: application-load-balancer
...
     regions:
     - region: us-east-1
       subnet: internal (vpc0)
```
Comment (robfletcher) [inline keel-api/src/test/resources/examples/alb-example.yml]: Generally moving `vpcName` from the root of the spec into the `locations` and `subnet` to each individual region.

File: keel-api/src/test/resources/examples/cluster-example.yml (hunk 2 above)
Comment (asher) [reply keel-api/src/test/resources/examples/cluster-example.yml]: Can this replace the need for `subnet` in the regions map?

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/Location.kt (hunk 3 above)
Comment (asher) [reply keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/Location.kt]: This seems wrong, since a `Subnet` always includes a `vpcId`.

File: keel-api/src/test/resources/examples/cluster-example.yml (hunk 2 above)
Comment (robfletcher) [reply keel-api/src/test/resources/examples/cluster-example.yml]: No, I think a VPC has many subnets as I understand it

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/Location.kt (hunk 3 above)
Comment (robfletcher) [reply keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/Location.kt]: An alternative approach I considered was to look up the VPC based on the subnet when needed. Because we have subnets with no "purpose" property (which is what this `subnet` property maps to) I wasn't sure that would cover edge cases. Then again, I guess this doesn't either.

File: keel-api/src/test/resources/examples/cluster-example.yml (hunk 2 above)
Comment (asher) [reply keel-api/src/test/resources/examples/cluster-example.yml]: A VPC can have many subnets but I don't think that prevents us from attempting to provide an opinionated default. In the absence of either a subnet or vpc, we could attempt to default to a subnet of "internal (vpc0)" which will correlate to a vpcId if it exists. Given just a `vpcName`, we could attempt to find a subnet name matching `"internal ($vpcName)"`.

I think it makes more sense not to require `vpcName` or `subnet`, with both derived by default from a subnet with the purpose of `internal (vpc0)` if present. If you need to deploy elsewhere, only specifying a different subnet should always be sufficient. There are subnets without a listed purpose but I don't think those are properly supported by clouddriver anyways, they need to be tagged first.

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/Location.kt (hunk 3 above)
Comment (asher) [reply keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/Location.kt]: I'm 95% sure we can ignore subnets without a purpose for now.

File: keel-api/src/test/resources/examples/cluster-example.yml (hunk 2 above)
Comment (robfletcher) [reply keel-api/src/test/resources/examples/cluster-example.yml]: That all makes sense. I guess my only question is whether that reflects a Netflix-only opinion of network configuration. Can/should we assume that's also valid for OSS users? I guess we could make those defaults configurable down the road.

Comment (asher) [reply keel-api/src/test/resources/examples/cluster-example.yml]: I think Spinnaker is opinionated enough that a "purpose" is required for use in OSS, but not about its contents. What do you think of Spring configuration properties for potential defaults like this that aren't easily resolved by "all the things available", like AZ's?

Comment (robfletcher) [reply keel-api/src/test/resources/examples/cluster-example.yml]: Yes, that's where I was heading. Although, I feel like I'd like to do that as a separate piece of work.

Comment (luispollo) [reply keel-api/src/test/resources/examples/cluster-example.yml]: BTW, I couldn't find any information in the OSS docs about how VPCs and subnets are supposed to be setup, what tags are expected to exist on those resources for Spinnaker to find them (including the naming conventions above), etc. The closest thing I found was [this](https://www.spinnaker.io/setup/install/providers/aws/aws-ec2/) but it only discusses setting up the accounts.

File: keel-api/src/test/resources/examples/alb-example.yml (hunk 4 above)
Comment (luispollo) [reply keel-api/src/test/resources/examples/alb-example.yml]: Do we foresee a case where a user would deploy the same thing in subnets with different purposes in different regions (say `internal (vpc0)` in `us-east-1` and `external (vpc0)` in `us-west-2`)? Seems unlikely...

Comment (luispollo) [reply keel-api/src/test/resources/examples/alb-example.yml]: In fact, we already do that for load balancers in the UI with the guided setup:
image

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/Location.kt (hunk 3 above)
Comment (erikmunson) [reply keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/Location.kt]: how does this play with security groups not needing a subnet?

Comment (robfletcher) [reply keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/Location.kt]: They can just specify the VPC name or have it defaulted

File: keel-api/src/test/resources/examples/alb-example.yml (hunk 4 above)
Comment (robfletcher) [reply keel-api/src/test/resources/examples/alb-example.yml]: You mean `subnet` could be moved up from the individual region to sit next to `vpcName`? What do you think @asher ?

Comment (asher) [reply keel-api/src/test/resources/examples/alb-example.yml]: This seems right. The resolved subnet might differ across both regions and individual AZ's, but all should map back to the same named purpose.

Comment (erikmunson) [reply keel-api/src/test/resources/examples/alb-example.yml]: don't have an informed opinion on whether subnet should be regional or not in these specs, but worth thinking about whether abstracting up to `internal` vs. `external` like we do in Luis's UI shots should really be at this level or at another level that can automatically resolve all the appropriate subnets for each region. Those two threads of conversation might be separate.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/Locatable.kt [hunk 5]
```diff
@@ -26,5 +26,13 @@ interface Locatable<T : RegionSpec> : ResourceSpec {
...
+  /**
+   * If not specified here, this should be derived from [subnet] or use a default VPC name.
+   */
+  val vpcName: String?,
+  /**
+   * If not specified here, this should be derived from a default subnet purpose using [vpcName].
+   */
+  val subnet: String?,
   val regions: Set<T>
 )
```
Comment (robfletcher) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/Locatable.kt]: VPC name and subnet (purpose) are now both here. That means

1. every region must use a VPC with the same name.
2. every region must use a subnet with the same purpose.
3. annoyingly, the `subnet` property is present even for `SecurityGroupSpec` where it will always be `null` (or at least ignored).

File: keel-api/src/test/resources/examples/alb-example.yml (hunk 4 above)
Comment (luispollo) [reply keel-api/src/test/resources/examples/alb-example.yml]: nit: See, @asher even said it himself above: what we're naming in the spec is basically the purpose of the subnet. 😃 Anyway, I don't have an issue with keeping the subnet names in the spec (as opposed to the more abstract "purpose"), but in that case I'd suggest at least renaming the field to `subnetName` for consistency, or removing the `Name` suffix from `vpcName`.

File: keel-api/src/test/resources/examples/cluster-example.yml (hunk 2 above)
Comment (luispollo) [reply keel-api/src/test/resources/examples/cluster-example.yml]: For future reference (thanks @asher!): https://docs.armory.io/spinnaker-install-admin-guides/aws-subnets/#configuring-subnets

File: keel-api/src/test/resources/examples/alb-example.yml (hunk 4 above)
Comment (luispollo) [reply keel-api/src/test/resources/examples/alb-example.yml]: P.S. @asher Just enlightened me that the value we expect in that `subnet` field is actually the value of the `purpose` metadata tag attached to the resource. So my last point above is moot. Wondering if it would be best to name the property `subnetPurpose` for clarity, though...

Comment (robfletcher) [reply keel-api/src/test/resources/examples/alb-example.yml]: If you hit the CloudDriver endpoint you'll see that subnets don't have a "name" as such. The closest thing is the "purpose". But yeah, I'd be down with dropping `Name` from `vpcName`.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/Locatable.kt [hunk 6]
```diff
@@ -20,11 +20,31 @@ package com.netflix.spinnaker.keel.api
...
+   */
+  val subnet: String?,
+  override val regions: Set<RegionSpec>
+) : Locations<RegionSpec>
+
+data class SimpleLocations(
+  override val accountName: String,
+  override val vpcName: String?,
+  override val regions: Set<String>
+) : Locations<String>
```
Comment (robfletcher) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/Locatable.kt]: Instead of one type of `Locations` and 2 types of `RegionSpec` there are now 2 types of `Locations`, one of which just has regions as a `Set<String>` (so no more custom Jackson handling).

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/Locatable.kt [hunk 7]
```diff
@@ -20,11 +20,32 @@ package com.netflix.spinnaker.keel.api
...
+   */
+  val subnet: String?,
+  override val regions: Set<SubnetAwareRegionSpec>
+) : Locations<SubnetAwareRegionSpec>
+
+data class SimpleLocations(
+  override val accountName: String,
+  override val vpcName: String?,
+  override val regions: Set<SimpleRegionSpec>
+) : Locations<SimpleRegionSpec>
```
Comment (robfletcher) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/Locatable.kt]: Instead of one type of `Locations` and 2 types of `RegionSpec` there are now 2 types of `Locations`. This leads to more consistent YAML between resource types that specify zones and those that do not.

File: keel-api/src/test/resources/examples/security-group-example.yml [hunk 8]
```diff
@@ -10,13 +10,13 @@ spec:
     detail: ec2v1
   locations:
     accountName: test
+    vpcName: vpc0
     regions:
-      - us-west-2
-      - us-east-1
-  vpcName: vpc0
+    - name: us-west-2
+    - name: us-east-1
```
Comment (robfletcher) [inline keel-api/src/test/resources/examples/security-group-example.yml]: This is a change to how regions are represented in security groups. But… it means they look the same as in other resources when those other resources are using default AZs. I think this is a good thing.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/RegionSpec.kt [hunk 9]
```diff
@@ -18,5 +18,17 @@
 package com.netflix.spinnaker.keel.api
 
 interface RegionSpec {
-  val region: String
+  val name: String
```
Comment (robfletcher) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/RegionSpec.kt]: Changed this because it makes the YAML and code representations make more sense.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/Locatable.kt [hunk 10]
```diff
@@ -18,13 +18,34 @@
 package com.netflix.spinnaker.keel.api
 
 /**
- * An object which is located in an account and region
+ * An resource spec which is located in an account and one or more regions.
```
Comment (luispollo) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/Locatable.kt]: nit: "*A* resource spec..."

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resolvers/NetworkResolver.kt [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+import com.netflix.spinnaker.keel.api.ec2.ApplicationLoadBalancerSpec
+import com.netflix.spinnaker.keel.api.ec2.ClassicLoadBalancerSpec
+import com.netflix.spinnaker.keel.api.ec2.ClusterSpec
+import com.netflix.spinnaker.keel.ec2.SPINNAKER_EC2_API_V1
+import com.netflix.spinnaker.keel.plugin.Resolver
+import org.springframework.stereotype.Component
+
+abstract class NetworkResolver<T : Locatable<*>> : Resolver<T> {
+
+  protected fun SubnetAwareLocations.withResolvedNetwork(): SubnetAwareLocations {
```
Comment (luispollo) [inline keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resolvers/NetworkResolver.kt]: Just curious: is it a common pattern to have a class whose sole purpose is to declare extensions for a different class like you did here? It wasn't initially clear to me that `NetworkResolver` types were being used anywhere in the code until I went back and noticed the `withResolvedNetwork` methods and the overrides to `invoke` again. 
Also, it looks like all 3 sub-types of `NetworkResolver` have the exact same override for `invoke` and the only difference is they have a different `supportedKind` but I couldn't spot where that was used either. I guess in general I was just wondering if we could have just added the code to resolve locations to the base `ResourceSpec` class (although I do realize not all resources are "locatable", but maybe in the `Locatable` interface then?) to simplify things.

Comment (robfletcher) [reply keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resolvers/NetworkResolver.kt]: It's because `Resolver<T>` extends the functional type `(Resource<T>) -> Resource<T>` which specifies the `invoke` method. That's an operator function. If an object has `invoke` you can use it like a function: e.g. `myThing("foo")`.
